File: src/main/java/org/checkerframework/checker/mungo/typestate/graph/states/State.java
package org.checkerframework.checker.mungo.typestate.graph.states;

import org.checkerframework.checker.mungo.typestate.ast.TMethodNode;
import org.checkerframework.checker.mungo.typestate.ast.TStateNode;
import org.checkerframework.checker.mungo.typestate.graph.states.AbstractState;
import org.checkerframework.checker.nullness.qual.Nullable;

public class State extends AbstractState<TStateNode, TMethodNode> {

  public String name;

  protected State(String name) {
    super(null);
    this.name = name;
  }

  public State(@Nullable TStateNode node) {
    super(node);
    this.name = node == null ? "unknown" : node.name;
  }

  @Override
  public String toString() {
    return "State{name=" + name + ",node=" + node + "}";
  }
}

File: src/main/kotlin/org/checkerframework/checker/mungo/typestate/graph/states/State.kt
package org.checkerframework.checker.mungo.typestate.graph.states

import org.checkerframework.checker.mungo.typestate.ast.TMethodNode
import org.checkerframework.checker.mungo.typestate.ast.TStateNode

open class State : AbstractState<TStateNode> {
  var name: String

  protected constructor(name: String) : super(null) {
    this.name = name
  }

  constructor(node: TStateNode?) : super(node) {
    name = (if (node == null) "unknown" else node.name)!!
  }

  val transitions: MutableMap<TMethodNode, AbstractState<*>> = HashMap()

  open fun addTransition(transition: TMethodNode, destination: AbstractState<*>) {
    transitions[transition] = destination
  }

  override fun toString(): String {
    return "State{name=$name}"
  }
}
